packages/dds/shared-object-base/src/serializer.ts

```diff
@@ -58,7 +59,103 @@ export interface IFluidSerializer {
 }
 
 /**
- * Data Store serializer implementation
+ * Use for readonly view of Json compatible data that can also contain {@link @fluidframework/core-interfaces#IFluidHandle}s.
+ *
+ * To be "Json compatible" requires that the data is acyclic.
+ *
+ * Note that this does not robustly forbid non json comparable data via type checking,
+ * but instead mostly restricts access to the json compatible subset.
+ *
+ * See also {@link isFluidHandle}.
+ *
+ * Warning: Objects with a key "type" and value "__fluid_handle__" will not round trip correctly through `FluidSerializer`
+ * (due to decodeValue's use of isSerializedHandle causing them to be wrongly parsed as handles) and should be avoided.
+ * @public
+ */
+export type FluidSerializableReadOnly =
+	| IFluidHandle
+	| string
+	| number
+	| boolean
+	// eslint-disable-next-line @rushstack/no-new-null
+	| null
+	| readonly FluidSerializableReadOnly[]
+	| { readonly [P in string]?: FluidSerializableReadOnly };
+
+/**
+ * Check if a value in {@link FluidSerializableReadOnly} data is an {@link @fluidframework/core-interfaces#IFluidHandle}.
+ *
+ * Warning: Non-IFluidHandle objects with a key "IFluidHandle" will cause an assert.
+ * Prior to version 2.0.0-internal.7.4.0, they were unconditionally interpreted as IFluidHandle.
+ *
+ * @remarks
+ * Its possible to have serializable data that looks like an {@link @fluidframework/core-interfaces#IFluidHandle} but isn't one.
+ * This function cannot distinguish these from real IFluidHandles in all cases: changes to the IFluidHandle would be required to ensure this is always possible.
+ *
+ * For now this is a best effort implementation.
+ *
+ * @privateRemarks
+ * Any of the following changes to IFluidHandle would solve the ambiguity problem:
+ *
+ * - The "IFluidHandle" property could be required to be a cyclic reference back to the parent.
+ *
+ * - IFluidHandle could be made into a class which could be detected with `instanceof`.
+ *
+ * - IFluidHandle could be identified using a dedicated symbol.
+ *
+ * @public
+ */
+export function isFluidHandle(value: undefined | FluidSerializableReadOnly): value is IFluidHandle {
+	if (typeof value !== "object" || value === null) {
+		return false;
+	}
+	const partialHandle = value as Partial<IFluidHandle>;
+	const innerHandle = partialHandle.IFluidHandle;
+	if (innerHandle === undefined) {
+		return false;
+	}
+
+	// At this point, the value is known to be an object with an IFluidHandle.
+	// This function will heuristically validate that object really is an IFluidHandle.
+	// If the object fails this check, it will assert to detect problematic data that would have incorrectly been treated as a handle in previous versions of the serializer.
+
+	// Since json compatible data shouldn't have methods, and IFluidHandle requires one, use that as a redundant check:
+	const getMember = (value as Partial<IFluidHandle>).get;
```

Comment:
Since `isFluidHandle` is called repeatedly/recursively on a critical path (op processing), I'm wondering if we need to deliberately optimize it. Putting this comment here because maybe we put this under the `if (value !== innerHandle)` check so it (likely) never actually has to run.

But the comment appllies in general, but the happy path of the function looks pretty lean.  Do we have any perf benchmarks that cover this code that we could observe before/after?  I think we do but not sure where or how to intepret them.